File: service/admin_service.py
import datetime
from time import time

import bcrypt
import jwt

from model.admin import Admin
from service.config_service import ConfigService
from utils import secret


class AdminService:
    def __init__(self):
        pass

    @staticmethod
    def init(identifier: str, password: str, vertex_endpoint: str, vertex_display_name: str, vertex_description: str) -> dict:
        if Admin.select().exists():
            raise Exception("You are not allowed to perform this action")

        admin = Admin(
            identifier=identifier,
            password=bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode(),
            creator=identifier
        )

        admin.save()

        ConfigService.set_vertex_endpoint(vertex_endpoint)
        ConfigService.set_vertex_display_name(vertex_display_name)
        ConfigService.set_vertex_description(vertex_description)

        return {
            "id": admin.get_id(),
            "identifier": admin.identifier,
            "creator": admin.creator,
            "created_at": admin.created_at,
            "updated_at": admin.updated_at
        }

    @staticmethod
    def get_token(identifier: str, password: str) -> dict:
        admin = Admin.select().where(Admin.identifier == identifier).get_or_none()

        if not admin:
            raise Exception("Invalid identifier and password combination")

        if not bcrypt.checkpw(password.encode(), admin.password.encode()):
            raise Exception("Invalid identifier and password combination")

        vertex_endpoint = ConfigService.get_vertex_endpoint()
        jwt_signing_key = ConfigService.get_jwt_signing_key()

        return {
            "token": jwt.encode({
                "sub": admin.identifier,
                "type": "admin",
                "iat": int(time()),
                "iss": vertex_endpoint,
                "aud": vertex_endpoint,
            }, algorithm="HS256", key=jwt_signing_key),
        }

    @staticmethod
    def get(identifier):
        admin = Admin.select().where(Admin.identifier == identifier).get_or_none()

        if not admin:
            raise Exception(f"Admin {identifier} not found")

        return {
            "identifier": admin.identifier,
            "creator": admin.creator,
            "created_at": admin.created_at,
            "updated_at": admin.updated_at
        }

    @staticmethod
    def change_password(identifier: str, password: str) -> dict:
        result = Admin.update(
            password = bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode(),
            updated_at=datetime.datetime.now(tz=datetime.timezone.utc)
        ).where(Admin.identifier == identifier).execute()

        if result == 0:
            raise Exception(f"Admin {identifier} not found")

        return {
            "identifier": identifier
        }

    @staticmethod
    def add(identifier: str, creator: str) -> dict:
        if Admin.select().where(Admin.identifier == identifier).exists():
            raise Exception(f"Admin {identifier} already exists")

        password = secret.generate(16)

        admin = Admin(
            identifier=identifier,
            password=bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode(),
            creator=creator,
        )

        admin.save()

        return {
            "id": admin.get_id(),
            "identifier": admin.identifier,
            "password": password,
            "creator": admin.creator,
            "created_at": admin.created_at,
            "updated_at": admin.updated_at
        }

    @staticmethod
    def fetch(page=0, size=50) -> list[dict]:
        result = []

        for admin in Admin.select().paginate(page, size):
            result.append({
                "id": admin.get_id(),
                "identifier": admin.identifier,
                "creator": admin.creator,
                "created_at": admin.created_at,
                "updated_at": admin.updated_at
            })

        return result

    @staticmethod
    def delete(identifier: str) -> dict:
        count = Admin.delete().where(Admin.identifier == identifier).execute()

        if count == 0:
            raise Exception(f"Admin {identifier} not found")

        return {
            "identifier": identifier,
        }

    @staticmethod
    def reset_password(identifier: str) -> dict:
        password = secret.generate(16)
        AdminService.change_password(identifier, password)
        return {
            "identifier": identifier,
            "password": password
        }

File: internal/app/vertex/admin/admin_service.go
package admin

import (
	"errors"
	"fmt"
	"github.com/evernetproto/evernet/internal/app/vertex/config"
	"github.com/evernetproto/evernet/internal/pkg/auth"
	"golang.org/x/crypto/bcrypt"
	"gorm.io/gorm"
)

type Service struct {
	db            *gorm.DB
	configService *config.Service
}

func NewService(db *gorm.DB, configService *config.Service) *Service {
	return &Service{db: db, configService: configService}
}

func (s *Service) Init(request *InitRequest) (*Admin, error) {
	var count int64
	result := s.db.Model(&Admin{}).Count(&count)

	if result.Error != nil {
		return nil, result.Error
	}

	if count != 0 {
		return nil, fmt.Errorf("you are not allowed to perform this action")
	}

	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(request.Password), bcrypt.DefaultCost)

	if err != nil {
		return nil, err
	}

	admin := &Admin{
		Identifier: request.Identifier,
		Password:   string(hashedPassword),
		Creator:    request.Identifier,
	}

	result = s.db.Create(admin)

	if result.Error != nil {
		return nil, result.Error
	}

	err = s.configService.SetVertexEndpoint(request.VertexEndpoint)

	if err != nil {
		return nil, err
	}

	err = s.configService.SetVertexIdentifier(request.VertexIdentifier)

	if err != nil {
		return nil, err
	}

	err = s.configService.SetVertexDisplayName(request.VertexDisplayName)

	if err != nil {
		return nil, err
	}

	err = s.configService.SetVertexDescription(request.VertexDescription)

	if err != nil {
		return nil, err
	}

	return admin, nil
}

func (s *Service) GetToken(request *TokenRequest) (*TokenResponse, error) {
	admin := &Admin{}

	result := s.db.Where(&Admin{Identifier: request.Identifier}).First(admin)

	if result.Error != nil {
		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
			return nil, fmt.Errorf("invalid identifier and password combination")
		}

		return nil, result.Error
	}

	err := bcrypt.CompareHashAndPassword([]byte(admin.Password), []byte(request.Password))

	if err != nil {
		return nil, fmt.Errorf("invalid identifier and password combination")
	}

	vertexEndpoint, err := s.configService.GetVertexEndpoint()

	if err != nil {
		return nil, err
	}

	jwtSigningKey, err := s.configService.GetJwtSigningKey()

	if err != nil {
		return nil, err
	}

	token, err := auth.GenerateAdminToken(admin.Identifier, vertexEndpoint, jwtSigningKey)

	if err != nil {
		return nil, err
	}

	return &TokenResponse{Token: token}, nil
}
